test-harness/modules/mocker/endpoints/generic.py

```diff
@@ -0,0 +1,276 @@
+    """Endpoint base class, from which all Endpoints must inherit
+
+       This class represents common behaviour shared across all endpoints,
+       no matter the function or repository flavour (ee/open).
+
+       Ever endpoint must by default serve GOOD/expected data, and only after
+       changing it's state using it's methods, it may start serving something
+       else and/or simulate error conditions.
+
+       The state of the endpoint may be changed by tests/fixtures by executing
+       Mocker's .send_command() method which in turn redirect the call to the
+       correct endpoint call. For the sake of simplicity it is assumed that each
+       such method will have well-defined interface:
+        def do_something(self, aux_data=None):
+            return result
+
+        `aux_data` is a python dictionary that must provide all data required
+            by function to execute. It can be None if such data is not required
+        `result` can be anything that makes sense in particular function's case.
+    """
+    _context = None
+    _httpd_thread = None
+    _httpd = None
+
+    def __init__(self, endpoint_id):
+        """Initialize new Endpoint object
+
+        Args:
+            endpoint_id (str): ID of the endpoint that it should identify itself
+                with
+        """
+        initial_data = {"always_bork": False,
+                        "endpoint_id": endpoint_id,
+                        }
+        self._context = EndpointContext(initial_data)
+
+    @property
+    def id(self):
+        """Return ID of the endpoint"""
+        return self._context.data['endpoint_id']
+
+    def start(self):
+        """Start endpoint's threaded httpd server"""
+        log.debug("Starting endpoint `%s`", self.id)
+        self._httpd_thread.start()
+
+    def stop(self):
+        """Perform cleanup of the endpoint threads
+
+        This method should be used right before destroying the Endpoint object.
+        It takes care of stopping internal httpd server.
+        """
+        log.debug("Stopping endpoint `%s`", self.id)
+        self._httpd.shutdown()
+        self._httpd_thread.join()
+
+    def reset(self, aux_data=None):
+        """Reset endpoint to the default/good state
+
+        Args:
+            aux_data (dict): unused, present only to satisfy the endpoint's
+                method interface. See class description for details.
+        """
+        del aux_data
+        log.debug("Resetting endpoint `%s`", self.id)
+        # Locking is not really needed here as it is atomic op anyway,
+        # but let's be consistent
+        with self._context.lock:
+            self._context.data['always_bork'] = False
+
+    def always_bork(self, aux_data=None):
+        """Make endpoint always respond with an error
+
+        Args:
+            aux_data (dict): unused, present only to satisfy the endpoint's
+                method interface. See class description for details.
+        """
+        del aux_data
```

Comment:
Today I Learned: use del with no second parameter to be explicit about non-use of an argument.